### ISSUE TITLE
[IdentityMapStrategyDetector] [CanMapUsingMapToSameTypeRule] Add support for deep copy instead of shallow copy when the type is the same via attribute.

### ISSUE THREAD

**Issue by sanelli** [sanelli/Mappa#14]

**Comment by sanelli** [comment]
### Cursor implementation plan

Uploaded plan file: `issue_14_identity_deep_copy_9e3dc5d9.plan.md`

---
---
name: Issue 14 Identity Deep Copy
overview: "Implement GitHub issue #14: add `IdentityMapDeepCopy` to `MappaSettings` with ShallowCopy (default), DeepCopy (`MemberwiseClone`), and NestedDeepCopy (clone shell + recursively map fields), wired through the generator identity strategy, with tests, samples, docs, and two version bumps from `10.1.0-alpha.32`."
todos:
  - id: "1"
    content: 1. Mappa — add IdentityMapDeepCopySetting enum, extend MappaSettingsAttribute, update Mappa/README.md
    status: completed
  - id: "2"
    content: 2. Bump MappaVersion.targets to alpha.33 and run .\Scripts\RebuildUponVersionChange.ps1
    status: completed
  - id: "3"
    content: "3. Mappa.Generator — wire settings; add UseIdentityMapStrategyDetector; update identity strategy/detector/builder; remove TODO #14"
    status: completed
  - id: "4"
    content: 4. Mappa.Generator.Tests — add DeepCopy, NestedDeepCopy (class and struct), primitive guard, editorconfig-only, class-over-editorconfig, and method-over-class integration tests
    status: completed
  - id: "5"
    content: 5. Bump MappaVersion.targets to alpha.34 and run .\Scripts\RebuildUponVersionChange.ps1
    status: completed
  - id: "6"
    content: 6. Mappa.Samples — add IdentityMapDeepCopyMapper (class and struct models) and update Mappa.Samples/README.md
    status: completed
  - id: "7"
    content: 7. Mappa.Samples.Tests — add ReferenceEquals unit tests for ShallowCopy (same), DeepCopy/NestedDeepCopy (not same), and nested-field sharing
    status: completed
  - id: "8"
    content: 8. Mappa.Samples.Aot — add runner exercising shallow/deep/nested maps and register in Program.cs
    status: completed
  - id: "9"
    content: 9. Documentation — update mappa-attributes.md, tutorial.md, mappa-generator-algorithm.md, features.md, Mappa.Generator/README.md
    status: completed
  - id: "10"
    content: 10. Run .\Scripts\RunTestsAndReportCoverage.ps1 and fix StyleCop/Sonar violations
    status: completed
isProject: false
---

# Issue #14: Identity Map Deep Copy

## Goal

Implement [GitHub issue #14](https://github.com/sanelli/Mappa/issues/14): when the identity strategy maps a type to **itself** (`T → T`), allow `[MappaSettings]` to control whether the result is a **shallow reference**, a **MemberwiseClone shell**, or a **MemberwiseClone shell with recursively mapped fields**.

Remove the existing TODO in [`IdentityMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/IdentityMapStrategyDetector.cs) (line 51).

**Prerequisite:** current version is `10.1.0-alpha.32` in [`MappaVersion.targets`](MappaVersion.targets). Follow the same settings pipeline pattern used by [`EnumToEnumMapSetting`](Mappa/EnumToEnumMapSetting.cs) (#18) and [`EnumStringMapSetting`](Mappa/EnumStringMapSetting.cs) (#15).

## Issue Requirements (from #14)

| Value | Behavior |
|-------|----------|
| `Undefined` | Inherit from parent scope / `.editorconfig`; does not override stacked settings |
| `ShallowCopy` | **Default** — return the original reference (current behavior) |
| `DeepCopy` | Create a new instance via `MemberwiseClone()`; nested references are **shared** |
| `NestedDeepCopy` | `MemberwiseClone()` plus recursively copy **every instance field** |

Additional constraints from the issue:
- Setting applies to **reference types** at the identity root; **primitive/value-only roots** (`int`, enums, etc.) always assign and ignore the setting.
- **Structs** are value types but may contain reference-type fields; **`NestedDeepCopy` must be supported for same-type struct roots** so those nested references are properly cloned.
- Nested logic must still handle **value-type fields that contain reference-type data** (e.g. a struct field nested inside a class).
- Generated code may use `Object.MemberwiseClone()` for **reference-type** roots; struct roots use struct copy plus field reassignment.

## Scope Boundaries

**In scope**
- New public enum + `MappaSettings` property
- Full settings stack: attribute → class/method → `.editorconfig`
- Identity detector/builder changes for same-type mappings via [`IdentityMapStrategyDetector`](Mappa.Generator/Algorithm/StrategyDetectors/IdentityMapStrategyDetector.cs)
- Generator integration tests, runtime sample + unit tests, AOT runner, documentation
- Two alpha version bumps + rebuilds

**Out of scope (unchanged behavior)**
- Same-type constructor-parameter identity path in [`ConstructorMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/ConstructorMapStrategyDetector.cs) (line 310) — continues to use bare `IdentityMapStrategy` (shallow pass-through); **do not modify this file for #14**
- Implicit conversions (`int → long`, `DateTime → DateTimeOffset`) — still assign/convert
- Boxing to `object` / `object?` — still assign
- Polymorphic identity override paths — unchanged
- New diagnostics (issue does not request any unless an unrecoverable edge case appears during implementation)

**Explicit special cases**

| Case | Behavior |
|------|----------|
| Primitive / enum root (`int → int`, `MyEnum → MyEnum`) | Always assign; ignore `IdentityMapDeepCopy` |
| Struct root (`MyStruct → MyStruct`) | **ShallowCopy / default:** struct assign (value copy; nested refs shared). **DeepCopy:** same as ShallowCopy at root (struct bitwise copy; nested refs still shared). **NestedDeepCopy:** struct copy (`var clone = source;`) then recursively map instance fields (including reference-type and nested struct fields) |
| Reference-type root (`Person → Person`, class/record) | **ShallowCopy:** return same reference. **DeepCopy:** `MemberwiseClone()` (new shell; nested refs shared). **NestedDeepCopy:** `MemberwiseClone()` + recursive field mapping |
| `string → string` | Always assign (immutable); ignore setting |
| Root `T[] → T[]` / collections | **DeepCopy:** cast `(T[])source.MemberwiseClone()` (new array, shared elements). **NestedDeepCopy:** bypass identity and resolve via container/array strategy so elements are mapped |
| `DeepCopy` / `NestedDeepCopy` on types without mutable fields | Still emit clone/copy; nested mode assigns only fields whose nested strategy produces code |

## Settings Design

### New enum: `IdentityMapDeepCopySetting`

Add [`Mappa/IdentityMapDeepCopySetting.cs`](Mappa/IdentityMapDeepCopySetting.cs) (mirror [`EnumToEnumMapSetting.cs`](Mappa/EnumToEnumMapSetting.cs)):

```csharp
public enum IdentityMapDeepCopySetting
{
    Undefined,
    ShallowCopy,
    DeepCopy,
    NestedDeepCopy,
}
```

### New `MappaSettings` property

Add to [`Mappa/Attributes/MappaSettingsAttribute.cs`](Mappa/Attributes/MappaSettingsAttribute.cs):

- **Property:** `IdentityMapDeepCopy` (type `IdentityMapDeepCopySetting`, default `Undefined`)
- **Effective default:** `ShallowCopy` when unset everywhere
- **`.editorconfig` key:** `mappa.identitymapdeepcopy`
- **Allowed values:** `ShallowCopy`, `DeepCopy`, `NestedDeepCopy`, `Undefined`

Document in [`Mappa/README.md`](Mappa/README.md) and [`Documentation/mappa-attributes.md`](Documentation/mappa-attributes.md) (property list + editorconfig table row).

## Generator Architecture

```mermaid
flowchart TD
    detect[IdentityMapStrategyDetector]
    sameType{SameTypeRule?}
    rootKind{Root kind}
    effective[Resolve effective IdentityMapDeepCopy]
    shallow[ShallowCopy: pass-through or struct assign]
    deep[DeepCopy: MemberwiseClone]
    nested[NestedDeepCopy: clone + field strategies]
    resolveFields[TypeMapIdentifierAlgorithm per field with identity disabled]
    build[IdentityMapStrategyBuilder emits code]

    detect --> sameType
    sameType -->|no| shallow
    sameType -->|yes| rootKind
    rootKind -->|primitive enum or string| shallow
    rootKind -->|struct| effective
    rootKind -->|class or record| effective
    effective -->|ShallowCopy| shallow
    effective -->|DeepCopy on struct| shallow
    effective -->|DeepCopy on class| deep
    effective -->|NestedDeepCopy| nested
    nested --> resolveFields
    shallow --> build
    deep --> build
    nested --> build
```

### 1. Settings pipeline (Generator)

Wire the new setting through the same touch points as `EnumToEnumMapSetting`:

| File | Change |
|------|--------|
| [`Mappa.Generator/Models/IMappaUserSettings.cs`](Mappa.Generator/Models/IMappaUserSettings.cs) | Add `IdentityMapDeepCopySetting IdentityMapDeepCopy { get; }` |
| [`Mappa.Generator/Models/MappaUserSettings.cs`](Mappa.Generator/Models/MappaUserSettings.cs) | `StackSetting<IdentityMapDeepCopySetting>`, ctor param, property, `Apply(...)` |
| [`Mappa.Generator/Models/MappaGlobalOptions.cs`](Mappa.Generator/Models/MappaGlobalOptions.cs) | Parse `mappa.identitymapdeepcopy`; default `ShallowCopy` |
| [`Mappa.Generator/Extensions/AttributeDataExtensions.cs`](Mappa.Generator/Extensions/AttributeDataExtensions.cs) | Read `nameof(MappaSettingsAttribute.IdentityMapDeepCopy)` |

Add helper (local static is fine):

```csharp
static IdentityMapDeepCopySetting GetEffective(IdentityMapDeepCopySetting value)
    => value is IdentityMapDeepCopySetting.Undefined
        ? IdentityMapDeepCopySetting.ShallowCopy
        : value;
```

### 2. Algorithm context flag for nested field resolution

Add `UseIdentityMapStrategyDetector` to [`MappaMapAlgorithmContextSettings.cs`](Mappa.Generator/Models/MappaMapAlgorithmContextSettings.cs) (same pattern as `UseConstructorMapStrategyDetector` / `UseNullableMapStrategyDetector`).

Update [`TypeMapIdentifierAlgorithm.cs`](Mappa.Generator/Algorithm/TypeMapIdentifierAlgorithm.cs) to skip `IdentityMapStrategyDetector` when the flag is `false`.

**Why:** `List<T> → List<T>` and `T[] → T[]` currently match identity first (shallow). NestedDeepCopy on fields/collections must fall through to [`ContainerMapStrategyDetector`](Mappa.Generator/Algorithm/StrategyDetectors/ContainerMapStrategyDetector.cs) / constructor strategies instead.

### 3. Extend `IdentityMapStrategy` model

Update [`IdentityMapStrategy.cs`](Mappa.Generator/Models/Strategies/IdentityMapStrategy.cs):

- Carry `IdentityMapDeepCopySetting` effective mode
- Carry `bool RequiresMemberwiseClone` (reference-type roots only)
- Carry `bool IsStructRoot` (struct same-type mapping)
- For `NestedDeepCopy` on classes/records/structs: `IReadOnlyList<(IFieldSymbol Field, MapStrategy Strategy)>` (or equivalent) resolved at detection time

Update `GetBuilder()` to return `new IdentityMapStrategyBuilder(this)`.

### 4. Detection changes

**[`IdentityMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/IdentityMapStrategyDetector.cs)**

When any identity rule matches:
1. If rule is **same-type** (`CanMapUsingMapToSameTypeRule()`):
   - **Primitive / enum roots:** build `IdentityMapStrategy` in **ShallowCopy** mode (assign); ignore setting
   - **`string`:** build in **ShallowCopy** mode (assign); ignore setting
   - **Struct roots:** read `context.MappaUserSettings.IdentityMapDeepCopy`, resolve effective setting
     - **ShallowCopy / DeepCopy:** struct assign (value copy; nested refs shared)
     - **NestedDeepCopy:** enumerate all instance fields on the struct (and nested struct fields as needed); resolve nested strategies with `UseIdentityMapStrategyDetector = false`; build strategy with field list
   - **Reference-type roots** (class/record, not array/collection/string): read setting and apply as below
     - **ShallowCopy:** pass-through
     - **DeepCopy:** `MemberwiseClone()`
     - **NestedDeepCopy** on composite reference types (class/record, not array/collection):
       - Enumerate **all instance fields** along the inheritance chain (`IFieldSymbol`, skip `static`/`const`)
       - For each field, create a derived [`MappaMapAlgorithmContext`](Mappa.Generator/Models/MappaMapAlgorithmContext.cs) for `fieldType → fieldType` with `UseIdentityMapStrategyDetector = false`
       - Run [`TypeMapIdentifierAlgorithm`](Mappa.Generator/Algorithm/TypeMapIdentifierAlgorithm.cs) to obtain nested strategies
       - If any nested resolution reports errors, abort identity detection (return `NoMapStrategy`)
     - **NestedDeepCopy** on **array/collection** same-type root: resolve strategy with identity disabled (expect container strategy)
     - Build `IdentityMapStrategy` with mode + nested strategies
2. Otherwise (implicit conversion, object boxing): build `IdentityMapStrategy` in **ShallowCopy** mode

**Not in scope:** [`ConstructorMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/ConstructorMapStrategyDetector.cs) (line 310) keeps its existing `new IdentityMapStrategy(...)` for single-parameter same-type constructors. `IdentityMapDeepCopy` applies only when the identity strategy is selected directly by `IdentityMapStrategyDetector` (root or nested type mappings), not via the constructor detector’s same-type parameter shortcut.

### 5. Code generation

**[`IdentityMapStrategyBuilder.cs`](Mappa.Generator/Builders/Strategies/IdentityMapStrategyBuilder.cs)**

| Mode | Generated pattern |
|------|-------------------|
| ShallowCopy (reference root) | `(source, string.Empty)` — unchanged |
| ShallowCopy (struct root) | `(source, string.Empty)` — struct assign via return |
| DeepCopy (reference root) | `T __tmp = (T)source.MemberwiseClone();` |
| DeepCopy (struct root) | struct assign (same as ShallowCopy; nested refs shared) |
| NestedDeepCopy (reference root) | `MemberwiseClone()` temp, then for each field: read `source.field`, run nested builder code, assign `clone.field = mapped;` |
| NestedDeepCopy (struct root) | `T __tmp = source;` then for each field: read `source.field`, run nested builder code, assign `__tmp.field = mapped;` |

Implementation notes:
- Use `context.NextTemporary()` for temps; push composite source name when generating nested field access (same pattern as [`PropertyMapStrategyBuilder`](Mappa.Generator/Builders/Strategies/PropertyMapStrategyBuilder.cs))
- Nullable reference roots: nullable strategy runs **before** identity; generated clone code operates on the non-null value
- Struct roots cannot call `MemberwiseClone()` from generated mapper code; use struct copy + mutable field reassignment for `NestedDeepCopy`
- Struct **fields inside** a cloned class: nested strategies handle struct→struct via field mapping when identity is disabled (including struct `NestedDeepCopy` when setting applies at that nested level)
- Remove TODO `[#14]`

## Tests ([`Mappa.Generator.Tests/IdentityStrategyIntegrationTests.cs`](Mappa.Generator.Tests/IdentityStrategyIntegrationTests.cs))

Add integration tests (happy + negative paths). Mirror the stacking test structure from [`EnumToEnumMapStrategyIntegrationTests`](Mappa.Generator.Tests/EnumToEnumMapStrategyIntegrationTests.cs) (`CanMapEnumToEnumWhenClassEnumToEnumMapSettingOverridesEditorConfig`, `CanMapEnumToEnumWhenMethodEnumToEnumMapSettingOverridesClassAndEditorConfig`).

### Core behavior

1. **Default shallow** — existing `string → string` tests remain valid (return `input`)
2. **DeepCopy class** — `[MappaSettings(IdentityMapDeepCopy = DeepCopy)]` on `Person → Person`; assert generated code contains `MemberwiseClone()` and does **not** return `input` directly
3. **NestedDeepCopy class** — type with reference + value fields; assert clone + field assignment statements
4. **NestedDeepCopy struct** — struct with a reference-type field; assert struct temp + field reassignment (no `MemberwiseClone()` on struct root)
5. **Value type primitive ignored** — `[MappaSettings(IdentityMapDeepCopy = DeepCopy)]` on `int → int`; still returns `input`
6. **NestedDeepCopy collection field** — class with `List<Child>`; assert new list construction (not identity pass-through on the list field)

### Settings stacking (required)

Use a simple reference-type same-type map (`Person → Person` or similar) and assert generated code to prove which effective setting won:

| Test | `.editorconfig` | Class `[MappaSettings]` | Method `[MappaSettings]` | Expected generated code |
|------|-----------------|-------------------------|--------------------------|-------------------------|
| **Editorconfig only** | `mappa.identitymapdeepcopy = DeepCopy` | *(none)* | *(none)* | `MemberwiseClone()` |
| **Class overrides editorconfig** | `DeepCopy` | `ShallowCopy` | *(none)* | returns `input` directly (no `MemberwiseClone()`) |
| **Method overrides class and editorconfig** | `DeepCopy` | `ShallowCopy` | `NestedDeepCopy` | clone + nested field assignments (not bare `input`, not shallow-only clone) |

Suggested test names (follow existing repo naming):

- `CanMapReferenceTypeToSameReferenceWhenIdentityMapDeepCopyIsSetInEditorConfig`
- `CanMapReferenceTypeToSameReferenceWhenClassIdentityMapDeepCopyOverridesEditorConfig`
- `CanMapReferenceTypeToSameReferenceWhenMethodIdentityMapDeepCopyOverridesClassAndEditorConfig`

Each stacking test must pass an `editorConfig` string to `RunMappaGeneratorAsync(sourceCode, editorConfig, ...)` and assert **syntax of the generated map method body**, not just absence of diagnostics — so the override under test is unambiguous.

Follow `HasSyntaxNodesCount(X)` / `HasNextSyntaxNode` chaining rules from workflow.

## Samples

Add [`Mappa.Samples/IdentityMapDeepCopyMapper.cs`](Mappa.Samples/IdentityMapDeepCopyMapper.cs) with separate map methods (or mappers) for each mode, using a reference-type model with at least one nested reference field, and a struct model with a reference-type field:

| Map method setting | Root `ReferenceEquals(source, result)` | Nested ref field `ReferenceEquals` |
|--------------------|----------------------------------------|-------------------------------------|
| **ShallowCopy** (default) | `true` | `true` (same nested instance) |
| **DeepCopy** | `false` (new shell) | `true` (nested ref still shared) |
| **NestedDeepCopy** (class root) | `false` | `false` (nested ref cloned) |
| **NestedDeepCopy** (struct root) | N/A (value type) | `false` for struct’s reference field vs source’s field |

Update [`Mappa.Samples/README.md`](Mappa.Samples/README.md).

### [`Mappa.Samples.Tests`](Mappa.Samples.Tests) — required runtime assertions

Add [`IdentityMapDeepCopyMapperUnitTest.cs`](Mappa.Samples.Tests/IdentityMapDeepCopyMapperUnitTest.cs) (or equivalent) with explicit `ReferenceEquals` checks:

- **ShallowCopy:** `Assert.True(ReferenceEquals(source, result))` and `Assert.True(ReferenceEquals(source.Nested, result.Nested))`
- **DeepCopy:** `Assert.False(ReferenceEquals(source, result))` and `Assert.True(ReferenceEquals(source.Nested, result.Nested))` (nested sharing proves deep-not-nested)
- **NestedDeepCopy (class):** `Assert.False(ReferenceEquals(source, result))` and `Assert.False(ReferenceEquals(source.Nested, result.Nested))`
- **NestedDeepCopy (struct):** `Assert.False(ReferenceEquals(source.Child, result.Child))` for the struct’s reference field (struct itself is a value copy)

Also cover **AOT** parity in [`Mappa.Samples.Aot`](Mappa.Samples.Aot) runner (same scenarios exercised at runtime).

## Documentation

| File | Update |
|------|--------|
| [`Documentation/mappa-attributes.md`](Documentation/mappa-attributes.md) | Property + editorconfig row |
| [`Documentation/tutorial.md`](Documentation/tutorial.md) | Short section with example |
| [`Documentation/mappa-generator-algorithm.md`](Documentation/mappa-generator-algorithm.md) | Identity strategy `_What_` / `_Notes_` (struct `NestedDeepCopy`, constructor-parameter path out of scope) |
| [`Documentation/features.md`](Documentation/features.md) | New row under MappaSettings table |
| [`Mappa.Generator/README.md`](Mappa.Generator/README.md) | Identity strategy section |

No new error codes expected; [`Documentation/error-codes.md`](Documentation/error-codes.md) unchanged unless implementation introduces a diagnostic.

## Versioning

| Step | Bump | When |
|------|------|------|
| After **Mappa** | `10.1.0-alpha.33` | Before Generator work |
| After **Mappa.Generator** (feature complete, tests pass) | `10.1.0-alpha.34` | Before Samples |

Run `.\Scripts\RebuildUponVersionChange.ps1` after each bump.

## Implementation Order (workflow)

### 1. Mappa — add `IdentityMapDeepCopySetting`, extend `MappaSettingsAttribute`, update `Mappa/README.md`

### 2. Bump `MappaVersion.targets` to `alpha.33` and run `.\Scripts\RebuildUponVersionChange.ps1`

### 3. Mappa.Generator — settings pipeline; `UseIdentityMapStrategyDetector`; extend identity strategy/detector/builder; remove TODO `#14` (no changes to `ConstructorMapStrategyDetector`)

### 4. Mappa.Generator.Tests — integration tests listed above, including explicit class-over-editorconfig and method-over-class stacking tests

### 5. Bump `MappaVersion.targets` to `alpha.34` and run `.\Scripts\RebuildUponVersionChange.ps1`

### 6. Mappa.Samples — `IdentityMapDeepCopyMapper` (class + struct models) + README

### 7. Mappa.Samples.Tests — `ReferenceEquals` unit tests for shallow vs deep vs nested (class and struct)

### 8. Mappa.Samples.Aot — runner + `Program.cs` registration

### 9. Documentation — attributes, tutorial, algorithm, features, generator README

### 10. Run `.\Scripts\RunTestsAndReportCoverage.ps1` and fix StyleCop/Sonar violations

## Key Files Reference

Existing shallow identity builder (baseline):

```17:20:Mappa.Generator/Builders/Strategies/IdentityMapStrategyBuilder.cs
    public (string VariableName, string Code) BuildSource(string source, MappaBuilderContext context, MappaGlobalOptions mappaGlobalOptions)
    {
        return (source, string.Empty);
    }
```

TODO insertion point:

```47:55:Mappa.Generator/Algorithm/StrategyDetectors/IdentityMapStrategyDetector.cs
        if (this.CanMapUsingMapToSameTypeRule()
            || this.CanMapUsingMapToObjectRule()
            || this.CanMapUsingImplicitConversion())
        {
            // TODO [#14] Add support for deep copy instead of shallow copy when the type is the same via attribute.
           mapStrategy = new IdentityMapStrategy(
                this.context.TargetType,
                this.context.SourceType);
        }
```

Settings pattern to mirror: [`EnumMapStrategyDetector.TryCreateEnumToEnumStrategy`](Mappa.Generator/Algorithm/StrategyDetectors/EnumMapStrategyDetector.cs) (reads `context.MappaUserSettings`, stores setting on strategy model).

## Risks and Mitigations

| Risk | Mitigation |
|------|------------|
| Same-type collections hit identity before container | `UseIdentityMapStrategyDetector = false` for nested/ collection NestedDeepCopy resolution |
| Constructor same-type parameter path ignores setting | Document as intentional out-of-scope; mapping via `IdentityMapStrategyDetector` only |
| Infinite recursion in nested field detection | Reuse existing constructor/nullable context guards; field strategies use normal detector chain with identity disabled |
| Init-only / readonly reference properties | Copy via **fields** after `MemberwiseClone`; document that nested mode requires assignable fields (same as BCL clone semantics) |
| Struct `NestedDeepCopy` at root | Use struct copy + field loop (not `MemberwiseClone`); enumerate all instance fields; resolve nested strategies with identity disabled |
| `MemberwiseClone` returns `object` | Cast to target type in generated code |

**Comment by sanelli** [comment]
### Cursor implementation plan

Uploaded plan file: `issue_14_identity_deep_copy_9e3dc5d9.plan.md`

---
---
name: Issue 14 Identity Deep Copy
overview: "Implement GitHub issue #14: add `IdentityMapDeepCopy` to `MappaSettings` with ShallowCopy (default), DeepCopy (`MemberwiseClone`), and NestedDeepCopy (clone shell + recursively map fields), wired through the generator identity strategy, with tests, samples, docs, and two version bumps from `10.1.0-alpha.32`."
todos:
  - id: "1"
    content: 1. Mappa — add IdentityMapDeepCopySetting enum, extend MappaSettingsAttribute, update Mappa/README.md
    status: completed
  - id: "2"
    content: 2. Bump MappaVersion.targets to alpha.33 and run .\Scripts\RebuildUponVersionChange.ps1
    status: completed
  - id: "3"
    content: "3. Mappa.Generator — wire settings; add UseIdentityMapStrategyDetector; update identity strategy/detector/builder; remove TODO #14"
    status: completed
  - id: "4"
    content: 4. Mappa.Generator.Tests — add DeepCopy, NestedDeepCopy (class and struct), primitive guard, editorconfig-only, class-over-editorconfig, and method-over-class integration tests
    status: completed
  - id: "5"
    content: 5. Bump MappaVersion.targets to alpha.34 and run .\Scripts\RebuildUponVersionChange.ps1
    status: completed
  - id: "6"
    content: 6. Mappa.Samples — add IdentityMapDeepCopyMapper (class and struct models) and update Mappa.Samples/README.md
    status: completed
  - id: "7"
    content: 7. Mappa.Samples.Tests — add ReferenceEquals unit tests for ShallowCopy (same), DeepCopy/NestedDeepCopy (not same), and nested-field sharing
    status: completed
  - id: "8"
    content: 8. Mappa.Samples.Aot — add runner exercising shallow/deep/nested maps and register in Program.cs
    status: completed
  - id: "9"
    content: 9. Documentation — update mappa-attributes.md, tutorial.md, mappa-generator-algorithm.md, features.md, Mappa.Generator/README.md
    status: completed
  - id: "10"
    content: 10. Run .\Scripts\RunTestsAndReportCoverage.ps1 and fix StyleCop/Sonar violations
    status: completed
isProject: false
---

# Issue #14: Identity Map Deep Copy

## Goal

Implement [GitHub issue #14](https://github.com/sanelli/Mappa/issues/14): when the identity strategy maps a type to **itself** (`T → T`), allow `[MappaSettings]` to control whether the result is a **shallow reference**, a **MemberwiseClone shell**, or a **MemberwiseClone shell with recursively mapped fields**.

Remove the existing TODO in [`IdentityMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/IdentityMapStrategyDetector.cs) (line 51).

**Prerequisite:** current version is `10.1.0-alpha.32` in [`MappaVersion.targets`](MappaVersion.targets). Follow the same settings pipeline pattern used by [`EnumToEnumMapSetting`](Mappa/EnumToEnumMapSetting.cs) (#18) and [`EnumStringMapSetting`](Mappa/EnumStringMapSetting.cs) (#15).

## Issue Requirements (from #14)

| Value | Behavior |
|-------|----------|
| `Undefined` | Inherit from parent scope / `.editorconfig`; does not override stacked settings |
| `ShallowCopy` | **Default** — return the original reference (current behavior) |
| `DeepCopy` | Create a new instance via `MemberwiseClone()`; nested references are **shared** |
| `NestedDeepCopy` | `MemberwiseClone()` plus recursively copy **every instance field** |

Additional constraints from the issue:
- Setting applies to **reference types** at the identity root; **primitive/value-only roots** (`int`, enums, etc.) always assign and ignore the setting.
- **Structs** are value types but may contain reference-type fields; **`NestedDeepCopy` must be supported for same-type struct roots** so those nested references are properly cloned.
- Nested logic must still handle **value-type fields that contain reference-type data** (e.g. a struct field nested inside a class).
- Generated code may use `Object.MemberwiseClone()` for **reference-type** roots; struct roots use struct copy plus field reassignment.

## Scope Boundaries

**In scope**
- New public enum + `MappaSettings` property
- Full settings stack: attribute → class/method → `.editorconfig`
- Identity detector/builder changes for same-type mappings via [`IdentityMapStrategyDetector`](Mappa.Generator/Algorithm/StrategyDetectors/IdentityMapStrategyDetector.cs)
- Generator integration tests, runtime sample + unit tests, AOT runner, documentation
- Two alpha version bumps + rebuilds

**Out of scope (unchanged behavior)**
- Same-type constructor-parameter identity path in [`ConstructorMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/ConstructorMapStrategyDetector.cs) (line 310) — continues to use bare `IdentityMapStrategy` (shallow pass-through); **do not modify this file for #14**
- Implicit conversions (`int → long`, `DateTime → DateTimeOffset`) — still assign/convert
- Boxing to `object` / `object?` — still assign
- Polymorphic identity override paths — unchanged
- New diagnostics (issue does not request any unless an unrecoverable edge case appears during implementation)

**Explicit special cases**

| Case | Behavior |
|------|----------|
| Primitive / enum root (`int → int`, `MyEnum → MyEnum`) | Always assign; ignore `IdentityMapDeepCopy` |
| Struct root (`MyStruct → MyStruct`) | **ShallowCopy / default:** struct assign (value copy; nested refs shared). **DeepCopy:** same as ShallowCopy at root (struct bitwise copy; nested refs still shared). **NestedDeepCopy:** struct copy (`var clone = source;`) then recursively map instance fields (including reference-type and nested struct fields) |
| Reference-type root (`Person → Person`, class/record) | **ShallowCopy:** return same reference. **DeepCopy:** `MemberwiseClone()` (new shell; nested refs shared). **NestedDeepCopy:** `MemberwiseClone()` + recursive field mapping |
| `string → string` | Always assign (immutable); ignore setting |
| Root `T[] → T[]` / collections | **DeepCopy:** cast `(T[])source.MemberwiseClone()` (new array, shared elements). **NestedDeepCopy:** bypass identity and resolve via container/array strategy so elements are mapped |
| `DeepCopy` / `NestedDeepCopy` on types without mutable fields | Still emit clone/copy; nested mode assigns only fields whose nested strategy produces code |

## Settings Design

### New enum: `IdentityMapDeepCopySetting`

Add [`Mappa/IdentityMapDeepCopySetting.cs`](Mappa/IdentityMapDeepCopySetting.cs) (mirror [`EnumToEnumMapSetting.cs`](Mappa/EnumToEnumMapSetting.cs)):

```csharp
public enum IdentityMapDeepCopySetting
{
    Undefined,
    ShallowCopy,
    DeepCopy,
    NestedDeepCopy,
}
```

### New `MappaSettings` property

Add to [`Mappa/Attributes/MappaSettingsAttribute.cs`](Mappa/Attributes/MappaSettingsAttribute.cs):

- **Property:** `IdentityMapDeepCopy` (type `IdentityMapDeepCopySetting`, default `Undefined`)
- **Effective default:** `ShallowCopy` when unset everywhere
- **`.editorconfig` key:** `mappa.identitymapdeepcopy`
- **Allowed values:** `ShallowCopy`, `DeepCopy`, `NestedDeepCopy`, `Undefined`

Document in [`Mappa/README.md`](Mappa/README.md) and [`Documentation/mappa-attributes.md`](Documentation/mappa-attributes.md) (property list + editorconfig table row).

## Generator Architecture

```mermaid
flowchart TD
    detect[IdentityMapStrategyDetector]
    sameType{SameTypeRule?}
    rootKind{Root kind}
    effective[Resolve effective IdentityMapDeepCopy]
    shallow[ShallowCopy: pass-through or struct assign]
    deep[DeepCopy: MemberwiseClone]
    nested[NestedDeepCopy: clone + field strategies]
    resolveFields[TypeMapIdentifierAlgorithm per field with identity disabled]
    build[IdentityMapStrategyBuilder emits code]

    detect --> sameType
    sameType -->|no| shallow
    sameType -->|yes| rootKind
    rootKind -->|primitive enum or string| shallow
    rootKind -->|struct| effective
    rootKind -->|class or record| effective
    effective -->|ShallowCopy| shallow
    effective -->|DeepCopy on struct| shallow
    effective -->|DeepCopy on class| deep
    effective -->|NestedDeepCopy| nested
    nested --> resolveFields
    shallow --> build
    deep --> build
    nested --> build
```

### 1. Settings pipeline (Generator)

Wire the new setting through the same touch points as `EnumToEnumMapSetting`:

| File | Change |
|------|--------|
| [`Mappa.Generator/Models/IMappaUserSettings.cs`](Mappa.Generator/Models/IMappaUserSettings.cs) | Add `IdentityMapDeepCopySetting IdentityMapDeepCopy { get; }` |
| [`Mappa.Generator/Models/MappaUserSettings.cs`](Mappa.Generator/Models/MappaUserSettings.cs) | `StackSetting<IdentityMapDeepCopySetting>`, ctor param, property, `Apply(...)` |
| [`Mappa.Generator/Models/MappaGlobalOptions.cs`](Mappa.Generator/Models/MappaGlobalOptions.cs) | Parse `mappa.identitymapdeepcopy`; default `ShallowCopy` |
| [`Mappa.Generator/Extensions/AttributeDataExtensions.cs`](Mappa.Generator/Extensions/AttributeDataExtensions.cs) | Read `nameof(MappaSettingsAttribute.IdentityMapDeepCopy)` |

Add helper (local static is fine):

```csharp
static IdentityMapDeepCopySetting GetEffective(IdentityMapDeepCopySetting value)
    => value is IdentityMapDeepCopySetting.Undefined
        ? IdentityMapDeepCopySetting.ShallowCopy
        : value;
```

### 2. Algorithm context flag for nested field resolution

Add `UseIdentityMapStrategyDetector` to [`MappaMapAlgorithmContextSettings.cs`](Mappa.Generator/Models/MappaMapAlgorithmContextSettings.cs) (same pattern as `UseConstructorMapStrategyDetector` / `UseNullableMapStrategyDetector`).

Update [`TypeMapIdentifierAlgorithm.cs`](Mappa.Generator/Algorithm/TypeMapIdentifierAlgorithm.cs) to skip `IdentityMapStrategyDetector` when the flag is `false`.

**Why:** `List<T> → List<T>` and `T[] → T[]` currently match identity first (shallow). NestedDeepCopy on fields/collections must fall through to [`ContainerMapStrategyDetector`](Mappa.Generator/Algorithm/StrategyDetectors/ContainerMapStrategyDetector.cs) / constructor strategies instead.

### 3. Extend `IdentityMapStrategy` model

Update [`IdentityMapStrategy.cs`](Mappa.Generator/Models/Strategies/IdentityMapStrategy.cs):

- Carry `IdentityMapDeepCopySetting` effective mode
- Carry `bool RequiresMemberwiseClone` (reference-type roots only)
- Carry `bool IsStructRoot` (struct same-type mapping)
- For `NestedDeepCopy` on classes/records/structs: `IReadOnlyList<(IFieldSymbol Field, MapStrategy Strategy)>` (or equivalent) resolved at detection time

Update `GetBuilder()` to return `new IdentityMapStrategyBuilder(this)`.

### 4. Detection changes

**[`IdentityMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/IdentityMapStrategyDetector.cs)**

When any identity rule matches:
1. If rule is **same-type** (`CanMapUsingMapToSameTypeRule()`):
   - **Primitive / enum roots:** build `IdentityMapStrategy` in **ShallowCopy** mode (assign); ignore setting
   - **`string`:** build in **ShallowCopy** mode (assign); ignore setting
   - **Struct roots:** read `context.MappaUserSettings.IdentityMapDeepCopy`, resolve effective setting
     - **ShallowCopy / DeepCopy:** struct assign (value copy; nested refs shared)
     - **NestedDeepCopy:** enumerate all instance fields on the struct (and nested struct fields as needed); resolve nested strategies with `UseIdentityMapStrategyDetector = false`; build strategy with field list
   - **Reference-type roots** (class/record, not array/collection/string): read setting and apply as below
     - **ShallowCopy:** pass-through
     - **DeepCopy:** `MemberwiseClone()`
     - **NestedDeepCopy** on composite reference types (class/record, not array/collection):
       - Enumerate **all instance fields** along the inheritance chain (`IFieldSymbol`, skip `static`/`const`)
       - For each field, create a derived [`MappaMapAlgorithmContext`](Mappa.Generator/Models/MappaMapAlgorithmContext.cs) for `fieldType → fieldType` with `UseIdentityMapStrategyDetector = false`
       - Run [`TypeMapIdentifierAlgorithm`](Mappa.Generator/Algorithm/TypeMapIdentifierAlgorithm.cs) to obtain nested strategies
       - If any nested resolution reports errors, abort identity detection (return `NoMapStrategy`)
     - **NestedDeepCopy** on **array/collection** same-type root: resolve strategy with identity disabled (expect container strategy)
     - Build `IdentityMapStrategy` with mode + nested strategies
2. Otherwise (implicit conversion, object boxing): build `IdentityMapStrategy` in **ShallowCopy** mode

**Not in scope:** [`ConstructorMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/ConstructorMapStrategyDetector.cs) (line 310) keeps its existing `new IdentityMapStrategy(...)` for single-parameter same-type constructors. `IdentityMapDeepCopy` applies only when the identity strategy is selected directly by `IdentityMapStrategyDetector` (root or nested type mappings), not via the constructor detector’s same-type parameter shortcut.

### 5. Code generation

**[`IdentityMapStrategyBuilder.cs`](Mappa.Generator/Builders/Strategies/IdentityMapStrategyBuilder.cs)**

| Mode | Generated pattern |
|------|-------------------|
| ShallowCopy (reference root) | `(source, string.Empty)` — unchanged |
| ShallowCopy (struct root) | `(source, string.Empty)` — struct assign via return |
| DeepCopy (reference root) | `T __tmp = (T)source.MemberwiseClone();` |
| DeepCopy (struct root) | struct assign (same as ShallowCopy; nested refs shared) |
| NestedDeepCopy (reference root) | `MemberwiseClone()` temp, then for each field: read `source.field`, run nested builder code, assign `clone.field = mapped;` |
| NestedDeepCopy (struct root) | `T __tmp = source;` then for each field: read `source.field`, run nested builder code, assign `__tmp.field = mapped;` |

Implementation notes:
- Use `context.NextTemporary()` for temps; push composite source name when generating nested field access (same pattern as [`PropertyMapStrategyBuilder`](Mappa.Generator/Builders/Strategies/PropertyMapStrategyBuilder.cs))
- Nullable reference roots: nullable strategy runs **before** identity; generated clone code operates on the non-null value
- Struct roots cannot call `MemberwiseClone()` from generated mapper code; use struct copy + mutable field reassignment for `NestedDeepCopy`
- Struct **fields inside** a cloned class: nested strategies handle struct→struct via field mapping when identity is disabled (including struct `NestedDeepCopy` when setting applies at that nested level)
- Remove TODO `[#14]`

## Tests ([`Mappa.Generator.Tests/IdentityStrategyIntegrationTests.cs`](Mappa.Generator.Tests/IdentityStrategyIntegrationTests.cs))

Add integration tests (happy + negative paths). Mirror the stacking test structure from [`EnumToEnumMapStrategyIntegrationTests`](Mappa.Generator.Tests/EnumToEnumMapStrategyIntegrationTests.cs) (`CanMapEnumToEnumWhenClassEnumToEnumMapSettingOverridesEditorConfig`, `CanMapEnumToEnumWhenMethodEnumToEnumMapSettingOverridesClassAndEditorConfig`).

### Core behavior

1. **Default shallow** — existing `string → string` tests remain valid (return `input`)
2. **DeepCopy class** — `[MappaSettings(IdentityMapDeepCopy = DeepCopy)]` on `Person → Person`; assert generated code contains `MemberwiseClone()` and does **not** return `input` directly
3. **NestedDeepCopy class** — type with reference + value fields; assert clone + field assignment statements
4. **NestedDeepCopy struct** — struct with a reference-type field; assert struct temp + field reassignment (no `MemberwiseClone()` on struct root)
5. **Value type primitive ignored** — `[MappaSettings(IdentityMapDeepCopy = DeepCopy)]` on `int → int`; still returns `input`
6. **NestedDeepCopy collection field** — class with `List<Child>`; assert new list construction (not identity pass-through on the list field)

### Settings stacking (required)

Use a simple reference-type same-type map (`Person → Person` or similar) and assert generated code to prove which effective setting won:

| Test | `.editorconfig` | Class `[MappaSettings]` | Method `[MappaSettings]` | Expected generated code |
|------|-----------------|-------------------------|--------------------------|-------------------------|
| **Editorconfig only** | `mappa.identitymapdeepcopy = DeepCopy` | *(none)* | *(none)* | `MemberwiseClone()` |
| **Class overrides editorconfig** | `DeepCopy` | `ShallowCopy` | *(none)* | returns `input` directly (no `MemberwiseClone()`) |
| **Method overrides class and editorconfig** | `DeepCopy` | `ShallowCopy` | `NestedDeepCopy` | clone + nested field assignments (not bare `input`, not shallow-only clone) |

Suggested test names (follow existing repo naming):

- `CanMapReferenceTypeToSameReferenceWhenIdentityMapDeepCopyIsSetInEditorConfig`
- `CanMapReferenceTypeToSameReferenceWhenClassIdentityMapDeepCopyOverridesEditorConfig`
- `CanMapReferenceTypeToSameReferenceWhenMethodIdentityMapDeepCopyOverridesClassAndEditorConfig`

Each stacking test must pass an `editorConfig` string to `RunMappaGeneratorAsync(sourceCode, editorConfig, ...)` and assert **syntax of the generated map method body**, not just absence of diagnostics — so the override under test is unambiguous.

Follow `HasSyntaxNodesCount(X)` / `HasNextSyntaxNode` chaining rules from workflow.

## Samples

Add [`Mappa.Samples/IdentityMapDeepCopyMapper.cs`](Mappa.Samples/IdentityMapDeepCopyMapper.cs) with separate map methods (or mappers) for each mode, using a reference-type model with at least one nested reference field, and a struct model with a reference-type field:

| Map method setting | Root `ReferenceEquals(source, result)` | Nested ref field `ReferenceEquals` |
|--------------------|----------------------------------------|-------------------------------------|
| **ShallowCopy** (default) | `true` | `true` (same nested instance) |
| **DeepCopy** | `false` (new shell) | `true` (nested ref still shared) |
| **NestedDeepCopy** (class root) | `false` | `false` (nested ref cloned) |
| **NestedDeepCopy** (struct root) | N/A (value type) | `false` for struct’s reference field vs source’s field |

Update [`Mappa.Samples/README.md`](Mappa.Samples/README.md).

### [`Mappa.Samples.Tests`](Mappa.Samples.Tests) — required runtime assertions

Add [`IdentityMapDeepCopyMapperUnitTest.cs`](Mappa.Samples.Tests/IdentityMapDeepCopyMapperUnitTest.cs) (or equivalent) with explicit `ReferenceEquals` checks:

- **ShallowCopy:** `Assert.True(ReferenceEquals(source, result))` and `Assert.True(ReferenceEquals(source.Nested, result.Nested))`
- **DeepCopy:** `Assert.False(ReferenceEquals(source, result))` and `Assert.True(ReferenceEquals(source.Nested, result.Nested))` (nested sharing proves deep-not-nested)
- **NestedDeepCopy (class):** `Assert.False(ReferenceEquals(source, result))` and `Assert.False(ReferenceEquals(source.Nested, result.Nested))`
- **NestedDeepCopy (struct):** `Assert.False(ReferenceEquals(source.Child, result.Child))` for the struct’s reference field (struct itself is a value copy)

Also cover **AOT** parity in [`Mappa.Samples.Aot`](Mappa.Samples.Aot) runner (same scenarios exercised at runtime).

## Documentation

| File | Update |
|------|--------|
| [`Documentation/mappa-attributes.md`](Documentation/mappa-attributes.md) | Property + editorconfig row |
| [`Documentation/tutorial.md`](Documentation/tutorial.md) | Short section with example |
| [`Documentation/mappa-generator-algorithm.md`](Documentation/mappa-generator-algorithm.md) | Identity strategy `_What_` / `_Notes_` (struct `NestedDeepCopy`, constructor-parameter path out of scope) |
| [`Documentation/features.md`](Documentation/features.md) | New row under MappaSettings table |
| [`Mappa.Generator/README.md`](Mappa.Generator/README.md) | Identity strategy section |

No new error codes expected; [`Documentation/error-codes.md`](Documentation/error-codes.md) unchanged unless implementation introduces a diagnostic.

## Versioning

| Step | Bump | When |
|------|------|------|
| After **Mappa** | `10.1.0-alpha.33` | Before Generator work |
| After **Mappa.Generator** (feature complete, tests pass) | `10.1.0-alpha.34` | Before Samples |

Run `.\Scripts\RebuildUponVersionChange.ps1` after each bump.

## Implementation Order (workflow)

### 1. Mappa — add `IdentityMapDeepCopySetting`, extend `MappaSettingsAttribute`, update `Mappa/README.md`

### 2. Bump `MappaVersion.targets` to `alpha.33` and run `.\Scripts\RebuildUponVersionChange.ps1`

### 3. Mappa.Generator — settings pipeline; `UseIdentityMapStrategyDetector`; extend identity strategy/detector/builder; remove TODO `#14` (no changes to `ConstructorMapStrategyDetector`)

### 4. Mappa.Generator.Tests — integration tests listed above, including explicit class-over-editorconfig and method-over-class stacking tests

### 5. Bump `MappaVersion.targets` to `alpha.34` and run `.\Scripts\RebuildUponVersionChange.ps1`

### 6. Mappa.Samples — `IdentityMapDeepCopyMapper` (class + struct models) + README

### 7. Mappa.Samples.Tests — `ReferenceEquals` unit tests for shallow vs deep vs nested (class and struct)

### 8. Mappa.Samples.Aot — runner + `Program.cs` registration

### 9. Documentation — attributes, tutorial, algorithm, features, generator README

### 10. Run `.\Scripts\RunTestsAndReportCoverage.ps1` and fix StyleCop/Sonar violations

## Key Files Reference

Existing shallow identity builder (baseline):

```17:20:Mappa.Generator/Builders/Strategies/IdentityMapStrategyBuilder.cs
    public (string VariableName, string Code) BuildSource(string source, MappaBuilderContext context, MappaGlobalOptions mappaGlobalOptions)
    {
        return (source, string.Empty);
    }
```

TODO insertion point:

```47:55:Mappa.Generator/Algorithm/StrategyDetectors/IdentityMapStrategyDetector.cs
        if (this.CanMapUsingMapToSameTypeRule()
            || this.CanMapUsingMapToObjectRule()
            || this.CanMapUsingImplicitConversion())
        {
            // TODO [#14] Add support for deep copy instead of shallow copy when the type is the same via attribute.
           mapStrategy = new IdentityMapStrategy(
                this.context.TargetType,
                this.context.SourceType);
        }
```

Settings pattern to mirror: [`EnumMapStrategyDetector.TryCreateEnumToEnumStrategy`](Mappa.Generator/Algorithm/StrategyDetectors/EnumMapStrategyDetector.cs) (reads `context.MappaUserSettings`, stores setting on strategy model).

## Risks and Mitigations

| Risk | Mitigation |
|------|------------|
| Same-type collections hit identity before container | `UseIdentityMapStrategyDetector = false` for nested/ collection NestedDeepCopy resolution |
| Constructor same-type parameter path ignores setting | Document as intentional out-of-scope; mapping via `IdentityMapStrategyDetector` only |
| Infinite recursion in nested field detection | Reuse existing constructor/nullable context guards; field strategies use normal detector chain with identity disabled |
| Init-only / readonly reference properties | Copy via **fields** after `MemberwiseClone`; document that nested mode requires assignable fields (same as BCL clone semantics) |
| Struct `NestedDeepCopy` at root | Use struct copy + field loop (not `MemberwiseClone`); enumerate all instance fields; resolve nested strategies with identity disabled |
| `MemberwiseClone` returns `object` | Cast to target type in generated code |
